### PR TITLE
TopFeat/results view top loci

### PR DESCRIPTION
**Pull request by jarvisraymond-uchicago**
Jira Ticket: [VADC-511](https://ctds-planx.atlassian.net/browse/VADC-511)

### New Features
This adds the Top Loci table to the ResultsPheWeb visualization. The functionality includes:
* Sorting
* Filtering
* Download of Manhattan Plot data in TSV format

Additionally the Results view has been organized into separate folders for the ResultsPheWeb, and the TopLociTable. Tests have been written and test data placed in the TestData folder.

### Bug Fixes
In some cases the Results view would crash the application instead of showing an error message if the data received from the API call was missing values. This has been fixed using the conditional operator, i.e. 
``` 
let results = data?.outputs?.parameters?.filter(
      (entry) => entry.name === 'pheweb_json_index',
    );
```
### Design

image

### Implementation
output

[VADC-511]: https://ctds-planx.atlassian.net/browse/VADC-511?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+      currentPage: 1,
+    });
+  };
+
+  const columns = [
+    {
+      title: 'Variant',
+      dataIndex: 'variant',
+      key: 'variant',
+      sorter: (a, b) => a.variant.localeCompare(b.variant),
```
Comment (pieterlukasse) [inline src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx:84]: we'll need a separate/custom sorter for this. Now we get the chromosomes sorted like this:
Screenshot 2023-06-21 at 19 33 25

1, 11, .. 2, 20, ...3, etc

File: src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+      children: [
+        {
+          title: (
+            <Input
+              placeholder='Search by Variant'
+              value={lociTableState.variantSearchTerm}
+              onChange={(event) => handleSearchTermChange(event, 'variant')}
```
Comment (pieterlukasse) [inline src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx:91]: here we also might need a custom search. E.g. now when we have:
> 1:192,575,969 G/ A (rs7535818)

and search for `192575969` we get no hits. The value should match though, regardless of the locale formatting used in the field itself.

File: src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx (hunk 1 above)
Comment (pieterlukasse) [reply src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx:84]: Similar problem in the p-value column: 

Screenshot 2023-06-21 at 19 44 41

File: src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      <Table
+        dataSource={filteredData}
+        columns={columns}
+        rowKey={(record) => record.pos}
+        onChange={handleTableChange}
+        pagination={{
+          current: lociTableState.currentPage,
+          defaultPageSize: 10,
+          showSizeChanger: true,
+          pageSizeOptions: ['10', '20', '30'],
```
Comment (pieterlukasse) [inline src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx:176]: I think ['10', '50', '100', '500', '1000'] can make this UI more useful.

**Review by pieterlukasse — COMMENTED** [review]
nice work @jarvisraymond-uchicago !

I tested locally and it works pretty well. I did find some minor sorting and filtering issues that would be good to address though. Please see comments.

File: src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx (hunk 3 above)
Comment (jarvisraymond-uchicago) [reply src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx:176]: Great idea. This has been updated to: 

```  pageSizeOptions: ['10', '50', '100','500','1000'], ```

File: src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx (hunk 2 above)
Comment (jarvisraymond-uchicago) [reply src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx:91]: Great idea. The search function has been omitted to strip commas from both the data at the key being searched and from the user input: 

```
const filterBySearchTerm = (data, key, searchTerm) => data.filter((obj) => obj[key]
  .toString()
  .toLowerCase()
  .replace(/,/g, '')
  .includes(searchTerm.toLowerCase().replace(/,/g, '')),
);
```

File: src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx (hunk 1 above)
Comment (jarvisraymond-uchicago) [reply src/Analysis/GWASResults/Views/Results/ResultsPheWeb/TopLociTable/TopLociTable.jsx]: The sorting algorithms have been updated. 

For the PVAL and AF the number is parsed as a number before doing the comparison, which should make numbers in scientific notation render in the correct order. 

For example: 
```
  sorter: (a, b) => Number(a.pval) - Number(b.pval),
  ```


For the variant column, the numbers are sorted first by the chrom value, and if those match, then the sorting is followed by the pos value. 
For example: 
```
      sorter: (a, b) => {
        const chromA = Number(a.variant.split(':')[0]);
        const chromB = Number(b.variant.split(':')[0]);
        if (chromA === chromB) {
          let posA = (a.variant.split(':')[1]);
          posA = posA.substring(0, posA.indexOf(' '));
          posA = Number(posA.replace(/,/g, ''));

          let posB = (b.variant.split(':')[1]);
          posB = posB.substring(0, posB.indexOf(' '));
          posB = Number(posB.replace(/,/g, ''));
          return posA - posB;
        }
        return chromA - chromB;
      },
      ```

**Review by pieterlukasse — APPROVED** [review]

**Pull request merged**